src/main/java/org/elasticsearch/node/Node.java

```diff
@@ -258,6 +258,11 @@ public Node start() {
         injector.getInstance(SearchService.class).start();
         injector.getInstance(MonitorService.class).start();
         injector.getInstance(RestController.class).start();
+
+        // TODO hack around circular dependecncies problems
+        injector.getInstance(AsyncShardFetch.Started.class).setReallocation(injector.getInstance(ClusterService.class), injector.getInstance(AllocationService.class));
```

Comment:
need to do this to get away from circular dep problems, but will move it to `GatewayAllocator` so it won't be those async fetches and encapsulated in the allocator itself